### ISSUE TITLE
Potential bug: Speed sensor loses tracking

### ISSUE THREAD

**Issue by MrHumanRebel** [danielnilsson9/bbs-fw#114]
Recently rode the bike with the latest 1.3.1 version for a long run, I noticed around every 10-15 minutes the speed sensor data is not shown properly for a few seconds. 
For example cruising at constant 35 km/h and sor a few moments firstly a lower speed is shown like 15-20 km/h and then a spike up to the 50s, then it is back to normal. 
Might be a sensor pulse filtering issue, but I'm just guessing at this point, I have no more information about his anomaly. 

The speed sensor and the magnet is located perfectly to each other, and working (and so blinking) perfectly fine.
This is with a BBS02B.

**Comment by danielnilsson9** [comment]
No idea. I think there is one other guy on a BBS02 who had similar issue too along with a lot of other issues. Haven't seen a this on my BBSHD or TSDZ2. I can leave the issue open but if this was a common firmware issue on the BBS02 I would have expected quite a few reports.

I own no bike with BBS02 so I have no experience of how it runs on that controller, support was added blindly by me and then tested by others.

**Comment by kb79000** [comment]
Hi, i did not notice this problem on my bbs02b

**Comment by silenec** [comment]
Yeah I did notice this on my last longer ride with road speed limited PAS levels. The motor randomly cuts power for no apparent reason and spins up again while the speed on the display goes up to 50s. Not very often though as you say so it is hard to diagnose.

**Comment by ronmicro** [comment]
> Yeah I did notice this on my last longer ride with road speed limited PAS levels. The motor randomly cuts power for no apparent reason and spins up again while the speed on the display goes up to 50s. Not very often though as you say so it is hard to diagnose.

My bbs02b with v1.3.1 has this same problem of random power cuts, it happens about 2 or 3 times in a 10 mile ride

**Comment by danielnilsson9** [comment]
Alright, definitely sounds like a firmware issue then. Was anyone of you running 1.2? Was the issue present there too or is it new in 1.3?

**Comment by MrHumanRebel** [comment]
> Alright, definitely sounds like a firmware issue then. Was anyone of you running 1.2? Was the issue present there too or is it new in 1.3?

Personally, I started using this firmware from 1.3.1 so I have no information regarding this.

**Comment by ronmicro** [comment]
I started from 1.3.1 too but i just tried version 1.2.0 on a long ride and it still happened a couple of times

**Comment by Larswagner1987** [comment]
> Ich habe auch mit 1.3.1 angefangen, aber ich habe gerade Version 1.2.0 auf einer langen Fahrt ausprobiert und es ist immer noch ein paar Mal passiert

that's a shame, do you have any idea what could be wrong, when does it happen, is there a trigger or did you notice something?

**Comment by Gonespa** [comment]
Hi,

I Have 1.2.0 too and noticed the same. I though this was related to the fact that i have max current set to 33 amps in a bbs02 and it was hitting some kind of extreme condition but I see it might be due to other thing. In my case, I noticed that behaviour when locking the rotor and demanding max current for a few seconds. I managed to see an error coming in the log.

**Comment by Larswagner1987** [comment]
I also suspect that it has something to do with the current, every time the motor is supposed to give current it happens, on the road it's just not a problem but as soon as there is an increase it starts as soon as he has to give more current then he interrupts 1-2 seconds and then the line comes back, I have a bbs02b motor 48v 750Watt

**Comment by dav0000000** [comment]
Are you guys using pas or throttle when it happens and if you are in SPORT, is it still in SPORT after the problem?

**Comment by Larswagner1987** [comment]
ich benutze nur pas und die normal , also kein Gas oder Sport

**Comment by Gonespa** [comment]
I have only noticed this when using the throttle. I am not 100% sure about the mode, but I would say thay the mode is not changing. 
If it happened I would probably not notice it since I have configured the "standard" mode as full power mode (and the one I use 99% of the time0 and "sport" mode as a tamed down one.

**Comment by Larswagner1987** [comment]
I have no gas in use, I tried again to generate the error and it only comes when the engine draws power, then the power goes away and comes back then it's fine and at some point it starts again, it's weird, it can  but just look at my display and see how the watt reading goes to 0 in time and then is full again.  must be a small thing, but where only in the log is there nothing in it for errors.

**Comment by danielnilsson9** [comment]
@Gonespa Probably not related to this issue unless others also have their current set that high. It's been reported before that it's not possible to set current above 25A for BBS02 because some other hardware protection is triggered outside the control of this firmware, that is probably what you are seeing. Set max current to 25A and it should work normally.

**Comment by Larswagner1987** [comment]
@danielnilsson9 I have mine set to 23a and it can be 25a, could it be because the engine had the misfires as a result???

**Comment by Gonespa** [comment]
> @Gonespa Probably not related to this issue unless others also have their current set that high. It's been reported before that it's not possible to set current above 25A for BBS02 because some other hardware protection is triggered outside the control of this firmware, that is probably what you are seeing. Set max current to 25A and it should work normally.

Agreed, I only mention it since the described behaviour looks exactly as I saw. I mean, behaves just like when the motor control micro returns an error: power disconnects completely for 1 or 2 seconds and the resets. Looks to me it is returning some kind of error. I was thinking, is it possible to save in permanent storage "the first error" returned, so that it can be retrieved like a car MCU does with the app?. 

Anyways, other than that extreme condition, I also noticed this a few times when using the bike in much less stressing conditions.

**Comment by Larswagner1987** [comment]
@danielnilsson9 The only thing I have done now is connect the laptop and start and then the error comes, but it only shows Motor target current changed to 0% for 1 second and then it goes back to normal, and once it says Failed to read voltage from motor controller, everything was fine so far.

**Comment by Larswagner1987** [comment]
@danielnilsson9 

I've tried everything now and it's a pity that this error is there, as soon as the laptop is connected it doesn't come up anymore as soon as it's not connected it comes up which is strange, the cables are new.  I tried different settings but it didn't work, I don't know what the problem is anymore, sometimes it's there immediately, sometimes it's only there after 4 km, then it doesn't work at all, it just breaks off and then the performance comes back after 1-2  Seconds, I've got the original bbs02 version back on it now, if you find the bug and then have a new version I'll try it.

**Comment by danielnilsson9** [comment]
I see, yes you are probably better off with the original firmware. These issues are not likely to be fixes as I don't have a bbs02 and are not doing any work on that motor. So unless someone else decides to do some investigation here it will probably stay this way.

**Comment by Larswagner1987** [comment]
@danielnilsson9 

ok shame, like I said, it's great except for the error, I can only send you a bbs02 controller or the original hex file, can you do something with it then?

**Comment by silenec** [comment]
I got back from another ride and I can say this bug appears way more often and / or is more noticeable when going uphill.
I can pretty much predict some situations where it happens. It almost always happens when approaching a hill.
- when road speed goes down and current goes up, the motor stops and starts again like 90% of the time
- it rarely happens when road speed goes up and current goes down (going downhill)
- it happens less often when going at constant speed in lower PAS levels
- it happens less often when road speed is higher (= more pulses from the speed sensor) even with stronger assist level
- it is probably the same problem I mentioned in https://github.com/danielnilsson9/bbs-fw/issues/105
- when going uphill in PAS5 (PAS + MaxSpeedPercent, config file in the link above), I got the motor to constantly skip even without throttle - previously I though it only happened with PAS + Throttle + MaxSpeedPercent but that is not the case
- it does not happen when using throttle only (no PAS)
I think we should test some bare minimum firmware - no temperature monitoring, no throttle, no keep current, no push walk, no ramping anything that is not necessary, maybe even no speed sensor if possible etc. and see when it stars working properly.

**Comment by Larswagner1987** [comment]
@silenec then try it and let me know, I definitely need the speed sensor and the shift sensor too, I don't need the rest like the gas, I hope that the error trigger is found, the original version is really not nice to use there  this version is perfect.  thank you already 👍

**Comment by danielnilsson9** [comment]
@Larswagner1987 Thanks but I have a BBS02 controller and firmware but diagnosing things that only occurs while riding on high power is hard with a controller on my desk.

@silenec That makes sense, I was never able to reproduce that issue, I have ridden > 100km since then using PAS and throttle on high power on my BBSHD but it's rock solid, I have not seen a single glitch. 

Weird thing is that there is not really any BBS02 specific code here in this firmware apart from pin assignments and motor MCU configuration. The motor MCU config was directly mirrored from the original BBS02 firmware. 

Are you seeing speed glitches on the display too?

**Comment by silenec** [comment]
Sometimes I see speed glitches on my display when it happens but sometimes I do not yet the motor still stops. It is weird. Maybe something in the NEC MCU does not like how the new FW behaves? Maybe there are some limits like maximum change of current / % of requested power per second or something and the limits might be different for BBSHD? And when something goes over the limit, NEC MCU "resets" and starts again? I have no idea honestly.

**Comment by danielnilsson9** [comment]
Ye maybe, I was thinking along these lines too as there is actually an error code reported from NEC MCU on BBS02 described here https://github.com/danielnilsson9/bbs-fw/wiki/NEC-Motor-Control-Protocol which I guessed was an overcurrent error but could be anything really. It's not exported to any error code sent to the display so you will never se an error when it occurs. Would be interesting to know if that error is set when this issue occurs.

NEC MCU shutting down doesn't explain the speed sensor reading issue though, since speed sensor reading is done by this firmware..

**Comment by silenec** [comment]
Now that I think about it... I put new chain on an older cassette once and the chain used to skip. The motor somehow detected that and every time the chain skipped, the motor did exactly this - stopped and spinned up again. Same thing when my rear tire slipped on ice in winter - motor stopped and started again. Not sure how that worked exactly. That was a few years ago on stock FW. Might it be the "chain skip protection" (if there is one) getting triggered by glitchy speed sensor values? Is the speed sensor reading reported to the NEC MCU or not? I guess not so this would not make sense then... hmm

**Comment by dav0000000** [comment]
I had speed sensor rise with a watchdog reset a while back. Only way I knew was SPORT mode always reset to STANDARD as the system restarted.

**Comment by Larswagner1987** [comment]
Hello

have you tested everything? a friend of mine thinks it could be something with the magnets, because when you drive and the power is gone, he took a closer look and noticed that it could be in the pas leagues, but there could also be incorrect information, the second could be there it's all mirrored what's wrong with the current values, since the BBs motor does a mini reset and then it's full again, or could it be that the memory/ram is too full and that's why it's happening?

**Comment by ronmicro** [comment]
I tried disabling the speed sensor in the config tool today but made no difference, still a lot of drops over a long ride

**Comment by danielnilsson9** [comment]
Maybe its RAM related anyways, its the only thing I can think of that is different between BBS02 and BBSHD, BBSHD has 4kb of ram, BBS02 has 2kb and there is only like 4 bytes of free ram left when compiling for BBS02. As per my current understanding that should be ok as everything is allocated upfront but maybe I have misunderstood something or there is a compiler bug causing these issues.

I will fix issue #86 and see if that helps, it's worth a shot. I'm also considering trying to compile with the Keil compiler and see if that makes any difference.

**Comment by MrHumanRebel** [comment]
> 

Thanks Daniel in advance for your hard work and patience debugging these stuff!

**Comment by Larswagner1987** [comment]
@danielnilsson9 Thanks also from me and my wife, you are doing a great job, you are the best. Thank you thank you thank you.

**Comment by silenec** [comment]
Meanwhile I went back to stock FW and the skipping is gone, no speed jumps on the lcd. So just to confirm, definitely not a hardware / sensor issue.

**Comment by silenec** [comment]
I used the biktrix one named CRR1094824E010015.3_170815.hex
https://biktrix.stonly.com/kb/guide/en/52v-programming-for-bbs02-motor-K2gUq2nWAX/Steps/1132765
It is a "52V" FW but it works fine with my 12S battery (minimum LVC is 38V, same as with the stock-stock FW before I flashed anything).

**Comment by danielnilsson9** [comment]
Thanks for testing! I figured it would not help while rereading all the comments in this thread. Must be something else. I'm gonna forward the error code from the NEC MCU to the display so we can know at least if that specific error is triggered and reported

**Comment by Larswagner1987** [comment]
Hello, have you already found the error? so far everything is going well except for error 09, I would be happy to hear that everything works without problems, the software is simply perfect and is what you want, I can only say thank you to Daniel for his great work on it.

**Comment by danielnilsson9** [comment]
> Hello, have you already found the error? so far everything is going well except for error 09, I would be happy to hear that everything works without problems, the software is simply perfect and is what you want, I can only say thank you to Daniel for his great work on it.

No I don't think it is resolved, see comment above by MrHumanRebel. The issue is still remains, fixing ram usage did not have any effect.

**Comment by Larswagner1987** [comment]
> > Hallo, hast du den Fehler schon gefunden? Bisher läuft bis auf Fehler 09 alles gut, es würde mich freuen zu hören, dass alles ohne Probleme funktioniert, die Software ist einfach perfekt und entspricht Ihren Wünschen, ich kann mich nur bei Daniel für die tolle Arbeit bedanken.
> 
> Nein, ich glaube nicht, dass es gelöst ist, siehe Kommentar oben von MrHumanRebel. Das Problem besteht immer noch, die Behebung der RAM-Nutzung hatte keine Auswirkung.

Okay👍

**Comment by MrHumanRebel** [comment]
Rode another 100+kms with my bike as I ride everyday to work and back. 
Funny story, some bug got into my eye and I looked down and saw the km/h jumping from 35 to 28.5 and back to 35. (the motor was going flawlessly as 28.5<35 so no dropout)
So that means the speed sensor spikes not only happen to the + direction to the 50s (then the motor stops obviosly 50>35), but also to the - direction.
I noticed this firstly today.

**Comment by ronmicro** [comment]
This might just be a coincidence but i changed my display from a DM03 to an 850C and didn't get any drops at all on my last 2 rides

**Comment by Larswagner1987** [comment]
how do you mean that no pot get???  I do not understand that .

**Comment by ronmicro** [comment]
Sorry what did you mean by no pot get? it didn't translate well.
Maybe some displays randomly send a value that makes it skip the switch case statement in process_bafang_display_write_pas()?

**Comment by ronmicro** [comment]
Changing that function a bit seems to have fixed it for my DM03 display, went 16 miles with no power cuts yesterday

**Comment by silenec** [comment]
[ronmicro](https://github.com/ronmicro) sounds good, is it still working for you? Can you share the modification?

**Comment by ronmicro** [comment]
Yes i've been on a few rides with no power cuts now.
This is in the extcom.c file line 721

```
static int16_t process_bafang_display_write_pas()
{
	if (msg_len < 4)
	{
		return KEEP;
	}

	switch (msgbuf[2])
	{
	case 0x00:
		app_set_assist_level(ASSIST_0);
		break;
	case 0x01:
		app_set_assist_level(ASSIST_1);
		break;
	case 0x0b:
		app_set_assist_level(ASSIST_2);
		break;
	case 0x0c:
		app_set_assist_level(ASSIST_3);
		break;
	case 0x0d:
		app_set_assist_level(ASSIST_4);
		break;
	case 0x02:
		app_set_assist_level(ASSIST_5);
		break;
	case 0x15:
		app_set_assist_level(ASSIST_6);
		break;
	case 0x16:
		app_set_assist_level(ASSIST_7);
		break;
	case 0x17:
		app_set_assist_level(ASSIST_8);
		break;
	case 0x03:
		app_set_assist_level(ASSIST_9);
		break;
	case 0x06:
		app_set_assist_level(ASSIST_PUSH);
		break;
	}

	return 4;
}
```

**Comment by Larswagner1987** [comment]
@ronmicro  

Can you tell us exactly what you did so that there are no more failures? I would be happy if it finally works without the failures, thanks in advance.

**Comment by ronmicro** [comment]
That is the edited function, if you compare the original i just prevented it from being able to default to ASSIST_0.
You could probably test this on v1.3.1 by just programming assist level 0 to be active.

**Comment by Larswagner1987** [comment]
> Das ist die bearbeitete Funktion. Wenn Sie das Original vergleichen, habe ich nur verhindert, dass sie standardmäßig auf ASSIST_0 eingestellt werden kann. Sie könnten dies wahrscheinlich auf Version 1.3.1 testen, indem Sie einfach die Unterstützungsstufe 0 so programmieren, dass sie aktiv ist.

So enter a value at level 0 that it is active or am I wrong there?

**Comment by aalm** [comment]
> So enter a value at level 0 that it is active or am I wrong there?

I think that would be rather non-optimal workaround, given you would have only one level of settings
working without the issue, and to fix that, rebuilding with the above change by ronmicro will be easier.

Btw., which display are you having this issue with?

**Comment by danielnilsson9** [comment]
@ronmicro Wow nice find! Thank you for investigating. Must be either some message sent that this firmware can't parse or some interference on the lines causing this. There is no checksum on this specific message...

I will add this fix for next version

**Comment by Larswagner1987** [comment]
where can you find the beta version or is it still uploaded??

**Comment by danielnilsson9** [comment]
@MrHumanRebel Thanks for testing, I will merge the changes. It's more of a workaround than an actual fix.

There are basically two things that are equally likely to cause these issues:

- An unknown command is sent by display which this firmware will not be able to handle and that will mess up the rest of the communication for a short period.
- Electrical interference. Sounds plausible based on reports that it's more likely to occur when high power is requested. Battery power cable routed along the display harness cable could make this more likely to occur.

@Larswagner1987 There are no precompiled beta versions, you will have to build from source or wait for release.

**Comment by Larswagner1987** [comment]
oh okay, then I'm curious when it will be published, or can someone create a file for me that I can upload it, I don't know how I can do it myself.

**Comment by silenec** [comment]
I flashed 1.4.0 and the workaround for this issue seems to work great - BBS02 is usable now.
Speed reading still goes up and down sometimes so speed limited PAS levels / street legal setup cannot be really used because of quite violent power drops and restarts. Other than that I am very happy with it. Seems to work perfectly with no (100km/h) speed limits set.

**Comment by Larswagner1987** [comment]
wie meinst du das ?? bitte mehr Details.

**Comment by silenec** [comment]
> wie meinst du das ?? bitte mehr Details.

https://endless-sphere.com/sphere/threads/bbs-fw-open-source-firmware-for-bbshd-bbs02-controller.117092/page-15#post-1770959

**Comment by Larswagner1987** [comment]
it turned out perfect thanks to everyone who helped 👍👍💪👍👍💪

**Comment by danielnilsson9** [comment]
@silenec So the speed limit kicks in when the speed on display goes up? That would mean that it is not just a communication error but rather a computation error if the computed speed is wrong too.

**Comment by Cabal83958** [comment]
BBS02 with 1.4.0. 
I have the weird readings too, both Watts and Speed readings. But the motor works perfectly so I don't really mind. I don't think it's electrical interference because I haven't changed the hardware at all and it didn't do that before.

**Comment by Larswagner1987** [comment]
I have my problems with me, it shows everything as it should be, no failure, no display problems

**Comment by Cabal83958** [comment]
> > Yes, it's limited to 60 km/h
> 
> Well and do you normally reach that speed? If not, then that's probably why it works fine for you.

only downhill and I haven't hit that speed yet with the 1.4.0

**Comment by Zensaition** [comment]
No fix for this issue at all is there no way to know what the system is doing with a data logger?

**Comment by Zensaition** [comment]
@Larswagner1987 what's your display? Mine is a dpc18 maybe screen shot your settings for the app?

**Comment by Larswagner1987** [comment]
I also installed the display on the second e-bike, including the Dz41, the settings on the display are all at the factory settings, I have version 1.4, which was newly released, installed on my controller, which one do you have???

**Comment by Zensaition** [comment]
@Larswagner1987 weird Idk I even flashed the firmware 2 times to make sure it stuck good? I have the STC15F2K60S2 on my bbs02b motor? im just confused if your dpc18 works... mine is from luna cycles? why is mine bugging out but works fine on stock firmware just goes crazy on this one!! ugh. I installed 1.2 like last year or when ever it was released did kinda the same thing. I have 1.4 installed might try again, Im 100% sure i did all the steps right. could be the speed sensor being old? i could replace that to see.

**Comment by Larswagner1987** [comment]
@Zensaition  I use the gear sensor in the same way as the brake, unfortunately I don't know how to upload a picture, I installed the software 1.4 for my wife and children and for me too, everything without errors, I'll try the data  to post on another site and then send you the link.

**Comment by Zensaition** [comment]
You can drag the screen shot or photo on the bottom bar of the box you type in its in grey writing. but that s good it works for you and your family lmao. yeah weird we have the same thing i guess besides the ebrake.

**Comment by Cabal83958** [comment]
Zensaition just use ur phone's gps for speed like the rest of us
why do you even care?

**Comment by Zensaition** [comment]
Cabal go be a crude phallic object elsewhere, I didnt ask you. I like accuracy and looking at my range.

**Comment by Cabal83958** [comment]
@Zensaition
Go fuck yourself you retarded piece of shit

**Comment by Cabal83958** [comment]
@Zensaition
oh and by the way the range works fine so your setup is just like you, shit 
pathetic asshole complaining and annoying everyone with your retarded bullshit you fucking loser and you DID ask me in PM on endlessphere so go fuck yourself again retarded asshole

**Comment by Cabal83958** [comment]
@Zensaition
oh and accuracy? you like accuracy heh? the fucking GPS is accurate not your stupid display you dumb fucking asshole

**Comment by Zensaition** [comment]
Man your attitude, really proves your childish hypocrisy, I could care less about your 1 brain cell why are you on here there are others with the same issue so i dont get your problem but obviously you are one damaged human being. I dont know who you are but I didnt message a guy like you. a Crude asshole, I could careless about

**Comment by silenec** [comment]
@danielnilsson9 I have managed to film a video with the latest FW version.
I use PAS levels 1-4 with my BBS02.
Global Max Current is set to 25A, all PAS levels use 100% max cadence and 100% max speed (100km/h).
PAS1 = 12% max current
PAS2 = 18%
PAS3 = 30%
PAS4 = 46%

PAS1 and PAS2 work just fine, no speed jumps on the LCD, no power drops
PAS3 works mostly fine - small speed jumps on the LCD when going uphill sometimes, no power drops
PAS4 works ok when going at fast and steady speed on flat terrain but shows massive speed jumps on the LCD when going uphill, motor randomly stops and restarts when going uphill when some unknown conditions are met (probably when the LCD speed jumps to maximum or minimum values). When the motor restart happens, it is very likely to happen again in a few seconds if I continue pedaling. Sometimes the motor does not restart but speed is jumping up and down on pretty much every hill.
Anything higher than PAS4 / 46% max current just means more frequent motor restarts.
Throttle only works just fine with 0-100% current - no speed jumps, no restarts even with full power.
No pedaling = no speed jumps.
PAS level with max speed set (like 25km/h or lower) = instant motor restarts once I reach that speed both on hills and flat terrain even with PAS1 / low current.  
Stock FW = no problems at all.

https://uploadnow.io/s/766c41d2-f4c6-4d41-a534-0d065ca68cca?o=t

**Comment by Larswagner1987** [comment]
Have you ever cleaned your magnet and is it positioned correctly, i.e. the distance???  For me and my wife it doesn't cause any problems, everything fits and works.

**Comment by silenec** [comment]
Yes magnet is fine, I checked it. LCD speed is perfect when using throttle only or with stock FW. So that pretty much rules it out I think.
However I wonder if having a full suspension bike could have something to do with the motor restarts. PAS actually activates sometimes without pedaling because the frame rotates a bit with the shock travel. I need to raise the Start Delay which is currently set at 90 degrees. 
So maybe it could also register as backpedaling when the shock travels in the opposite direction? That would not explain the speed jumps though. Anyway, Stop Delay is now 200ms - is that too low for a FS bike?
What else could cause the motor to stop? Is there any PAS minimum or maximum speed in the FW? Could it be disabled as a workaround for this issue? Motor temperature is <40C when this happens so that's probably not it.

**Comment by Larswagner1987** [comment]
@silenec I also have a full-suspension bike, it's not like that for me, no matter what I set, everything works as it should, I don't use gas, but I've had problems with that too and when I take the throttle off, I have no problems, pedaling backwards  The engine doesn't start, I don't know what's wrong with you, but for me it just gives me the reason to drive without the throttle and see if the problem is still there. Just try to adjust pas and deactivate the throttle and  let me know if it's still like that.  sorry my english is not that good and I use a translator,👍

**Comment by ronmicro** [comment]
What design of rear suspension do you have, is it possible for the magnet and sensor to shift out of alignment slightly as it  compresses?

**Comment by Larswagner1987** [comment]
@silenec It may be that it shifts if it is not properly fixed, of course you have to keep in mind that the position will then shift and it will then not display correctly.  Try driving normally without it bouncing and then you drive and jump a little so that the frame moves then you notice that there is a jump on the display.

**Comment by silenec** [comment]
@Larswagner1987 I have no problems with throttle. Motor does not start when pedaling backwards, it stops like it should. It just stops when it shouldn't.
@ronmicro I don't think so. Magnet is on the lower chainstay so it moves together with rear wheel. The bike: https://i.imgur.com/2vc1GRA.jpeg

Speed on the LCD jumps even on nice flat road without bumps in PAS4. That does not happen in PAS1 or PAS2 or with stock firmware.

**Comment by Larswagner1987** [comment]
then you have to use the originals again, as I said, no one has had any problems with the new software so far.

**Comment by silenec** [comment]
Well not really. I need lower voltage cutoff with my battery and I only get that with BBS-FW. It rides just fine in PAS1-3 and that is all I need most of the time (I reach 30 to 35+ km/h in PAS3). I can also use the throttle when I need more power. 
If you look at the original post, clearly I am not the only one with this problem. There are some more recent issues also https://github.com/danielnilsson9/bbs-fw/issues/195

**Comment by danielnilsson9** [comment]
@silenec Thank for detailed error description.

Power cut at high speed could be caused by calculated speed exceeding 100km/h... That means it's not a display communication issue like the other one mentioned in this thread.

That does however not make much sense when combined with only happening while using PAS.
You say "No pedaling = no speed jumps.". Do you have problems if you pedal while using throttle (on a throttle only level) or is it only in PAS?

**Comment by silenec** [comment]
@danielnilsson9 I think the LCD shows a bit averaged value (?) so the individual spikes could be much higher, maybe even exceeding 100km/h like you say or go near 0km/h when moving?

I had problems with pedaling + throttle + [1.3.1](https://github.com/danielnilsson9/bbs-fw/releases/tag/v1.3.1) with or without PAS but that got fixed with 1.4.0 - maybe this helped https://github.com/danielnilsson9/bbs-fw/commit/7ec3cc532df4142163f09f614c258aebc35d9c81
So yes, I can now pedal without PAS and use throttle at the same time with no problems and no speed jumps on the LCD. Even full throttle (limited to 80% in config) works just fine, no speed jumps, no power cuts.

It only happens with active PAS and only above ~30% current. It happened even with low current when I was on 1.3.1 and throttle + pedaling also was not working so it got much better with 1.4.0 but it is not completely fixed yet.

**Comment by ronmicro** [comment]
You could test disabling the speed sensor in the config tool, it looks like it skips the apply_speed_limit function then

**Comment by silenec** [comment]
I could try that when I find the time.
Correction: speed jumps are also present without PAS with heavy throttle + pedaling and I also see some at lower current PAS2 level when the motor spins up after a gear change. Much smaller speed jumps though and no power cuts in both cases.

**Comment by Zensaition** [comment]
> I could try that when I find the time. Correction: speed jumps are also present without PAS with heavy throttle + pedaling and I also see some at lower current PAS2 level when the motor spins up after a gear change. Much smaller speed jumps though and no power cuts in both cases.

Theses are exactly my issues and a few others. No matter what setting I turn off or on, it just jumps and breaks the sensors to go to 100+. Sometimes even power is limited or wont go past 5% on full throttle. I gotta hard power off to reset the system.... its gotta be maybe the firmware install process or the usb cables we are using. like if others arent complaining alot, which would be good to know how many downloads lol. to really know the ratio.... maybe mcu like different mhz? 

what I noticed is more then 600w and an uphill load with pas breaks it completely. so its gotta be conflict with all those together or just the watt meter to speed sensor or install wasnt perfect for some reason. I did it 4times now... step by step.

**Comment by silenec** [comment]
> You could test disabling the speed sensor in the config tool, it looks like it skips the apply_speed_limit function then

@danielnilsson9 @ronmicro 
Small update after a while - I have unchecked Speed Sensor in config tool and my problems are indeed resolved. I can finally use all of the power and I still see the speed on LCD even with speed sensor disabled so that is good. Only issue left is that I can't make the bike street legal with no speed limits but I can live with that for now.

**Comment by ronmicro** [comment]
i guess its missing a pulse, maybe the bafang firmware had some averaging to compensate for that.
you could try stacking a stronger neodymium magnet on top of your spoke magnet

**Comment by silenec** [comment]
Good idea, I will try that. The frame and spokes could be flexing a little bit when accelerating hard so that could explain why the speed reading gets worse under those conditions. I still get some inacurrate numbers and that does not happen with stock FW so you might be right.

**Comment by silenec** [comment]
I used two magnets on top of each other with the sensor really close to them. It made no difference at all - speed still jumps around so I continue to run with speed sensor disabled.

**Comment by MrHumanRebel** [comment]
Still an issue more than one year later, happens about 2-3 times on a 10km ride, on the BBS02 750W.
